sota-launcher/src/datatypes.rs

```diff
@@ -0,0 +1,234 @@
+use json;
+use reqwest;
+use serde::{self, Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashMap;
+use std::fmt::{self, Display, Formatter};
+use std::fs::File;
+use std::io::{self, Read};
+use std::result;
+use std::str::FromStr;
+use toml;
+use uuid;
+
+
+error_chain!{
+    foreign_links {
+        Http(reqwest::Error);
+        Io(io::Error);
+        Toml(toml::de::Error);
+        Uuid(uuid::ParseError);
+    }
+
+    errors {
+        Config(s: String) {
+            description("config error")
+            display("config error: '{}'", s)
+        }
+    }
+}
+
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum Environment {
+    CI,
+    QA,
+    Production,
+}
+
+impl FromStr for Environment {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s.to_lowercase().as_ref() {
+            "ci" => Ok(Environment::CI),
+            "qa" => Ok(Environment::QA),
+            "production" => Ok(Environment::Production),
+            _ => Err(format!("unknown environment: {}", s).into())
+        }
+    }
+}
+
+impl Display for Environment {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            Environment::CI => write!(f, "https://ci-app.atsgarage.com/api/v1"),
+            Environment::QA => write!(f, "https://qa-app.atsgarage.com/api/v1"),
+            Environment::Production => write!(f, "https://app.atsgarage.com/api/v1"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Environment {
+    fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
+        if let json::Value::String(ref s) = Deserialize::deserialize(de)? {
+            s.parse().map_err(|err| serde::de::Error::custom(format!("{}", err)))
+        } else {
+            Err(serde::de::Error::custom("expected string type for environment"))
+        }
+    }
+}
+
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UpdateTargets {
+    pub targets: HashMap<String, Update>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Update {
+    pub from: Option<UpdateTarget>,
+    pub to: UpdateTarget
+}
+
+#[allow(non_snake_case)]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UpdateTarget {
+    pub target: String,
+    pub targetLength: u64,
```

Comment:
`#[serde(rename = "targetLength")]` and then you can ditch `#[allow(non_snake_case)]`